### DISCUSSION TITLE
Why we need use `Controller` component for mui?

### DISCUSSION THREAD

**Discussion by valerii15298** [react-hook-form/react-hook-form#8189] [Q&A] [answered]
I am able to use `react-hook-form` with `mui` like this:
```ts
// ...some code before
const { ref: nameInputRef, ...nameInput } = register("name");
return  <TextField
          {...nameInput}
          inputRef={nameInputRef}
          defaultValue={"SomeDefValue"}
        />
```

But docs says you should use `<Controller/>` component. I had read that it is because `mui` components does not expose `ref` prop. But we can use `inputRef` instead.
My question is why we should use `<Controller/>`?
Does approach above has some disadvantages?

**Comment by Moshyfawn** [comment, marked as answer]
Hey, @valerii15298

You can actually just spread the `register` props into the MUI `TextField` input as they've started forwarding `ref`s instead of using a custom `ref` name since the docs were written. I can't remember the specific version MUI that started doing that, but if you search [the RHF discussions tab](https://github.com/react-hook-form/react-hook-form/discussions), you'll be able to find it.

Now, there's still an issue with the material's floating label and the uncontrolled inputs. In some cases, the label doesn't shrink using the `register` method _(uncontrolled component that is)_. Again, there's a discussion somewhere in [the RHF discussions tab](https://github.com/react-hook-form/react-hook-form/discussions) about the issue.

All that said, I'm using uncontrolled MUI `TextField` in my production app as I have the input labels always shrinked.
```jsx
<TextField
  InputLabelProps={{
    shrink: true
  }}
/>
```

At the end of the day, using `Controller` vs `register` is all about controlled vs uncontrolled inputs. The advantage of uncontrolled input is the source of truth for it's value is the input's value, not the outside state like `React.useState`, which means, no re-renders on value change.
